File: python/Two Pointers/trapping_rain_water.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        """
        Compute how much water can be trapped after raining using two pointers.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Edge case: empty array
        if not height:
            return 0
        # Initialize two pointers and their max heights
        l, r = 0, len(height) - 1
        leftMax, rightMax = height[l], height[r]
        res = 0
        # Continue until pointers meet
        while l < r:
            # If left max is smaller, process left side
            # This is because the water level is limited by the smaller max
            if leftMax < rightMax:
                l += 1 
                leftMax = max(leftMax, height[l])
                res += leftMax - height[l]
            # If right max is smaller or equal, process right side
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                res += rightMax - height[r]
        return res

# -------------------
# Test Cases
# -------------------
def test_trap():
    s = Solution()
    # Test case 1: Example
    height = [0,1,0,2,1,0,1,3,2,1,2,1]
    assert s.trap(height) == 6
    # Test case 2: No water
    height = [0,1,2,3,4]
    assert s.trap(height) == 0
    # Test case 3: All same height
    height = [2,2,2,2]
    assert s.trap(height) == 0
    # Test case 4: Single bar
    height = [4]
    assert s.trap(height) == 0
    # Test case 5: Empty input
    height = []
    assert s.trap(height) == 0
    # Test case 6: Multiple valleys
    height = [4,2,0,3,2,5]
    assert s.trap(height) == 9
    print("All test cases passed!")

if __name__ == "__main__":
    test_trap()

File: C++/Two Pointers/trapping_rain_water.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Calculate how much water can be trapped after raining
     * 
     * Approach:
     * - Use two pointers (left and right) from array ends
     * - Track maximum height seen from left and right
     * - Process smaller max height side first
     * - Water at position = min(leftMax, rightMax) - height[i]
     * 
     * Time: O(n) - single pass through array
     * Space: O(1) - only using pointers and variables
     */
    int trap(vector<int>& height) {
        // Edge case: empty array
        if (height.empty()) return 0;
        
        // Initialize pointers and result
        int left = 0;
        int right = height.size() - 1;
        int leftMax = height[left];
        int rightMax = height[right];
        int result = 0;
        
        // Process until pointers meet
        while (left < right) {
            // Process left side if left max is smaller
            if (leftMax < rightMax) {
                left++;
                leftMax = max(leftMax, height[left]);
                result += leftMax - height[left];
            }
            // Process right side if right max is smaller or equal
            else {
                right--;
                rightMax = max(rightMax, height[right]);
                result += rightMax - height[right];
            }
        }
        
        return result;
    }
};

/*
 * Detailed C++ Syntax Explanation:
 * 
 * 1. Vector Operations:
 *    height.empty()     // Check if vector is empty
 *    height.size()      // Get array length
 *    height[index]      // Access element at index
 * 
 * 2. Variable Declarations:
 *    int left = 0      // Left pointer starting at beginning
 *    int right = n-1   // Right pointer starting at end
 *    int result = 0    // Accumulate trapped water
 * 
 * 3. Standard Library:
 *    max(a, b)         // Returns larger of two values
 * 
 * 4. Edge Case:
 *    if (height.empty()) return 0;  // Early return for empty input
 * 
 * 5. Water Calculation:
 *    leftMax - height[left]    // Water trapped at left position
 *    rightMax - height[right]  // Water trapped at right position
 * 
 * 6. Two Pointer Movement:
 *    left++    // Move left pointer right
 *    right--   // Move right pointer left
 * 
 * Note: This solution achieves:
 * - O(n) time complexity by visiting each element at most once
 * - O(1) space complexity by using only variables
 */
